core/src/main/scala/org/apache/spark/sql/SnappyParser.scala

```diff
@@ -348,6 +348,42 @@ class SnappyParser(session: SnappySession)
       (d, s.asInstanceOf[Option[Duration]]))
   }
 
+  private def extractGroupingSet(
+      child: LogicalPlan,
+      aggregations: Seq[NamedExpression],
+      groupByExprs: Seq[Expression],
+      groupingSets: Seq[Seq[Expression]]): GroupingSets = {
+    val keyMap = groupByExprs.zipWithIndex.toMap
+    val bitmasks: Seq[Int] = groupingSets.map(set => set.foldLeft(0)((bitmap, col) => {
+      require(keyMap.contains(col), s"$col doesn't show up in the GROUP BY list")
+      bitmap | 1 << keyMap(col)
+    }))
+    GroupingSets(bitmasks, groupByExprs, child, aggregations)
+  }
+
+  protected final def groupingSetExpr: Rule1[Seq[Expression]] = rule {
+    '(' ~ ws ~ (expression + (',' ~ ws)) ~ ws ~ ')'  ~>  ((gs: Seq[Expression]) => gs) |
+        (expression + (',' ~ ws))  ~>  ((gs: Seq[Expression]) => gs) |
+        '(' ~ ws ~ ')'  ~>  (() => (Seq[Expression]()))
+  }
+
+  protected final def cubeRollUpGroupingSet: Rule1[(Seq[Seq[Expression]], Option[String])] = rule {
+    CUBE ~> (() => (Seq(Seq[Expression]()), Option("CUBE"))) |
+    ROLLUP ~> (() => (Seq(Seq[Expression]()),Option("ROLLUP"))) |
+    GROUPING ~ SETS ~ ('(' ~ ws ~ (groupingSetExpr ~ ws).+(',' ~ ws) ~ ws ~ ')' ~ ws)  ~>
+        ((gs: Seq[Seq[Expression]]) => (gs, Option("GROUPINGSETS")))
+  }
+
+  protected final def groupBy: Rule1[(Seq[Expression], Seq[Seq[Expression]],  Option[String])] = rule {
+    GROUP ~ BY ~ (expression + (',' ~ ws)) ~
+        (WITH ~ cubeRollUpGroupingSet).? ~>
+        ((groupingExpr: Seq[Expression], crgs: Option[(Seq[Seq[Expression]], Option[String])]) =>
+        { val emptyCubeRollupGrSet = (Seq(Seq[Expression]()), Option("")) // if cube, rollup, GrSet is not used
```

Comment:
Due to bugs in even recent IDEA in type inference (due to which it showed most of rules in red), I ended up declaring types for (?, +, *) which are normally (Option, Seq, Seq) respectively, to be Any and then an explicit cast in body. Perhaps you can do the same to remove those annoying error markers in IDEA.